src/bluesearch/database/topic_info.py

```diff
@@ -0,0 +1,116 @@
+from __future__ import annotations
+
+import copy
+import datetime
+import pathlib
+from dataclasses import dataclass, field
+from typing import Any
+
+import bluesearch
+from bluesearch.database.article import ArticleSource
+
+
+@dataclass
+class TopicInfo:
+    """The topic information extracted from a journal article.
+
+    For the spec see the following GitHub issue/comment:
+    https://github.com/BlueBrain/Search/issues/518#issuecomment-985525160
+    """
+
+    source: ArticleSource
+    path: str | pathlib.Path
+    element_in_file: int | None = None
+    article_topics: dict[str, list[str]] = field(init=False, default_factory=dict)
+    journal_topics: dict[str, list[str]] = field(init=False, default_factory=dict)
+    metadata: dict = field(default_factory=dict)
+
+    def __post_init__(self) -> None:
+        """Run the post-initialization."""
+        self.creation_date = datetime.datetime.now()
+
+    @staticmethod
+    def add_topics(mapping: dict[str, list[str]], kind: str, topics: list[str]) -> None:
+        """Add topics to a mapping with collection of topics.
+
+        Parameters
+        ----------
+        mapping
+            A mapping of the form kind -> list-of-topics that shall be
+            updated in-place. For example ``{"MeSH": ["topic 1", "topic 2"]}``.
+        kind
+            The topic kind. Corresponds to a key in ``mapping``.
+        topics
+            The topics to add. Corresponds to a value in ``mapping``.
+        """
+        updated_topics = mapping.get(kind, []) + topics
+        mapping[kind] = sorted(set(updated_topics))
+
+    def add_article_topics(self, kind: str, topics: list[str]) -> None:
+        """Add article topics.
+
+        Parameters
+        ----------
+        kind
+            The topic kind. For example "MeSH" or "MAG".
+        topics
+            A list of the topics to add.
+        """
+        self.add_topics(self.article_topics, kind, topics)
+
+    def add_journal_topics(self, kind: str, topics: list[str]) -> None:
+        """Add journal topics.
+
+        Parameters
+        ----------
+        kind
+            The topic kind. For example "MeSH" or "MAG".
+        topics
+            A list of the topics to add.
+        """
+        self.add_topics(self.journal_topics, kind, topics)
+
+    def json(self) -> dict:
+        """Convert the contents of this class to a structured dictionary.
+
+        Returns
+        -------
+        dict
+            The structure dictionary with all topic information.
+        """
+        metadata: dict[str, Any] = {
+            **copy.deepcopy(self.metadata),
+            "created-date": self.creation_date.strftime("%Y-%m-%d %H:%M:%S"),
+            "bbs-version": bluesearch.__version__,
+        }
+        if self.element_in_file is not None:
+            metadata["element_in_file"] = self.element_in_file
+
+        json = {
+            "source": self.source.value,
+            "path": str(self.path),
```

Comment:
Maybe we can keep the absolute path here ?

Update: I just saw that you are taking the absolute path when instantiating the TopicInfo into the entrypoint. Don't you think it could be worth to still take the absolute path here to be sure it is harmonized ?

Reply:
That's certainly possible. I left the responsibility for checking/resolving the path to the caller for two reasons
* We accept `pathlib.Path` and `str` types, so we'd have to convert back and forth
* If the path doesn't exist then resolving it could give unexpected results (from the [docs](https://docs.python.org/3/library/pathlib.html#pathlib.Path.resolve): "[if] the path is resolved as far as possible and any remainder is appended without checking whether it exists")

But it's not an absolute no-go. What do the others think?

Reply:
IMO both are okay as long as we clearly document the behavior somewhere (e.g. class docstring).

Reply:
I feel like Emilie that absolute paths are better, since we are then expecting to use this info when reading the file to the know which files should be loaded into the DB and which ones should not.

Reply:
Path resolution added here: 5145e04

Reply:
Thank you! :)